pkgs/tools/virtualization/nixos-nspawn/nixos-nspawn.py

```diff
@@ -0,0 +1,512 @@
+        else:
+            raise e
+
+
+def op_list_gens(args):
+    create_container(args.name, expect_imperative=True).list_generations()
+
+
+def op_rollback(args):
+    strategy = 'reload' if args.reload else ('restart' if args.restart else None)
+    create_container(args.name, expect_imperative=True).rollback(strategy=strategy)
+
+
+if __name__ == '__main__':
+    logging.basicConfig(
+        level=logging.INFO,
+        format='%(asctime)s %(levelname)-8s %(message)s'
+    )
+
+    parser = ArgumentParser()
+    parser.add_argument('--show-trace', action='store_true')
+    cmd = parser.add_subparsers(dest='subcmd')
+
+    # Options to `create` a new imperative container. Apart from the name,
+    # a config-file is accepted that's mostly equivalent to how declarative
+    # containers are built.
+    create = cmd.add_parser('create')
+    create.add_argument('name', help='Name of the container')
+    create.add_argument('config', help='Configuration file for the container')
+
+    # Options to `list` all containers (either pretty-printed or as json).
+    list = cmd.add_parser('list')
+    list.add_argument('--declarative', action='store_true')
+    list.add_argument('--imperative', action='store_true')
+    list.add_argument('--json', action='store_true')
+
+    # Options to `remove` a single, imperative container.
+    rm = cmd.add_parser('remove')
+    rm.add_argument('name', help='Name of the container')
+
+    # Options to `update` an imperative container by re-applying a Nix
+    # expression (as used by `create`) to it.
+    update = cmd.add_parser('update')
+    update.add_argument('name', help='Name of the container')
+    update.add_argument('--reload', help='Whether to reload the container', action='store_true')
+    update.add_argument('--restart', help='Whether to restart the container', action='store_true')
+    update.add_argument('--config', help='Configuration file to use to update the container')
+
+    # Wrapper for `nix-env --list-generations` on an imperative container.
+    list_gens = cmd.add_parser('list-generations')
+    list_gens.add_argument('name', help='Name of the container')
+
+    # Wrapper for `nix-env --rollback` on an imperative container.
+    rollback = cmd.add_parser('rollback')
+    rollback.add_argument('name', help='Name of the container')
+    rollback.add_argument('--reload', help='Whether to reload the container', action='store_true')
+    rollback.add_argument('--restart', help='Whether to restart the container', action='store_true')
+
+    args = parser.parse_args()
+    dst = args.subcmd
+    if not args.subcmd:
+        dst = "list"
+        args.imperative = True
+        args.declarative = True
+        args.json = False
+
+    cmds = {
+        'create': op_create,
+        'list': op_list,
+        'remove': op_remove,
+        'update': op_update,
+        'list-generations': op_list_gens,
+        'rollback': op_rollback,
+    }
+
+    try:
+        cmds[dst](args)
```

Comment:
There's a nice trick you can do with argparser to relate a function to a particular subcommand:

```python
# For each subcommand, function pair:
subcommand.set_defaults(handler=handling_func)
# After parsing, run the given handler
args = parser.parse_args()
handler = args.handler
handler(args)
```